Fix potential 'Floating Point Exception' with speculative execution

grospelliergilles opened the pull request:
With the current version of runtime and version `8.0-rc2`, there is a 'Floating Point Exception' (signal `SIGFPE`) when floating point exception are enabled (via `feenableexcept()`).

It seems to be due to speculative execution of the division in the check.

This patch remove the check and make sure we do not divide by zero.

Alternatively, because the result of this function is only to check if the fragmentation is high enough, we can rewrite the part of the code to remove all divide operation.

grospelliergilles committed:
Fix potential 'Floating Point Exception' with speculative execution

tannergooding commented:
Running with floating-point exceptions enabled is explicitly not supported, the behavior that occurs if it is enabled is undefined.

CC. @jkotas

grospelliergilles commented:
Thanks for your response.
I already had this kind of error in old versions of the runtime (#76257) and this was fixed.
In my application I do not have control on enabling floating point exception. And because it occurs in the garbage collector  I can not disable them in this part.
The fix is pretty simple and should not change the results. Do you think it could be merged ?

mangod9 commented:
@Maoni0 as well.

Maoni0 commented:
this is a different issue from the fix I made in https://github.com/dotnet/runtime/pull/76294/. if a speculative execution can result in SIGFPE (which seems quite dangerous to me), that means the fix I made would subject to the same problem. 

we should decide whether we want to support this speculative execution scenario at all - if we do, we should harden all the places this can happen instead of making one off fixes like this. 

I don't have a strong opinion on whether to support it or not, @jkotas, do you?

jkotas commented:
As I have said in https://github.com/dotnet/runtime/issues/76257#issuecomment-1260226375, .NET runtime and libraries are not compatible with enabled floating point exceptions. If you need floating point exceptions enabled for your C/C++ library, you need to disable around them around the calls to .NET runtime.

I do not think that one-off workarounds like the one proposed in this PR make sense.

grospelliergilles commented:
Thansk @jkotas for your response. It is difficult to disable FPE around this part of the code because it is called by the garbage collector and I do not know when it will be called. I will try to use `GC.RegisterForFullGCNotification()` to see if it works.

Alternatively, I think the code may be rewritten without using division because the result is only used in comparison. So instead of checking `a>b/c` we can rewrite as `a*c>b`. If you agree I can provide a patch to remove them if you want.

jkotas commented:
> So instead of checking a>b/c we can rewrite as a*c>b

`generation_allocator_efficiency` is called in number of places. I do not see how you can do this rewrite.

grospelliergilles committed:
Rewrite the calculation of 'generation_allocator_efficiency' and 'generation_unusable_fragmentation' to remove floating point usage.

- Rename 'generation_allocator_efficiency' to
  'generation_allocator_efficiency_percent' because now it returns a
  number between 0 and 100.
- Do not use 'generation_allocator_efficiency' in
  'generation_unusable_fragmentation' and do direct calculation.

grospelliergilles commented:
I have removed floating point usage in the two functions which can throw FPE.
Using only integer division there is no longer potential FPE.

grospelliergilles commented:
> @grospelliergilles please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @dotnet-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @dotnet-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @dotnet-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
@dotnet-policy-service agree

grospelliergilles commented:
@dotnet-policy-service agree

jkotas commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -5105,23 +5105,24 @@ size_t& generation_allocated_since_last_pin (generation* inst)
+    size_t free_list_allocated = generation_free_list_allocated (inst);
+    if (free_list_allocated==0)
+      return 0;
+    return (100 * free_list_allocated) / (free_list_allocated + free_obj_space);
```
Can this hit integer overflow?

grospelliergilles replied:
if `size_t` is 64bit, it can not overflow because `free_list_allocated` will never be greater than `2^60`.
I can use `uint64_t` if you prefer.

mangod9 replied:
@grospelliergilles, do you plan to update the PR with `uint64_t` and perhaps we can evaluate again.

grospelliergilles replied:
Thanks for you reply @mangod9. Yes I will update my PR if you want. I will do that this week

grospelliergilles committed:
Use 'uint64_t' instead of 'size_t' for computing generation fragmentation.

mangod9 commented:
@grospelliergilles, there appear to be build breaks on x86. @Maoni0 please review as well.

markples commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -5105,20 +5105,24 @@ size_t& generation_allocated_since_last_pin (generation* inst)
+    uint64_t free_list_space = generation_free_list_space (inst);
+    if (free_obj_space==0)
+      return 0;
+    return (free_obj_space + (free_obj_space * free_list_space) / (free_list_allocated + free_obj_space));
```
This could use a comment, either that it is based on rewriting of `(100 - efficiency)/100 * free_list_space` or `inefficiency * free_list_space`, being careful in both cases to minimize the impact of the integer division.

grospelliergilles replied:
Thanks for your report. I will make the required change and add the information your want.

Maoni0 replied:
I would really rather to not have to have the presence of FLT_EPSILON in the GC code at all if I can avoid it :) it's just another thing that someone who's not familiar with FP has to care about. 

but I agree with @markples that these should continue to return size_t, instead of uint64_t. I would also add a comment above `generation_allocator_efficiency_percent` to explain why this is written this way (to accommodate this exception specific to FP which is not a common knowledge AFAIK) instead of the previous way which seems more natural.

markples reviewed (commented):
I feel like there is some readability loss here.  Perhaps it can be addressed (see my specific remark on a comment that could help).  Also, I think it would help to limit the use of `uint64_t` to where the large values can occur.  `generation_allocator_efficiency_percent` can have an intermediate large value but returns a percentage.  `generation_unusable_fragmentation` also can have an intermediate large value but returns a size (in fact, this is the cause of the build break because the call site is appropriately expecting a `size_t`.

However, given this, I wonder if adding something like FLT_EPSILON to the denominator would be a clearer fix.  I thought about suggesting that `generation_unusable_fragmentation` continue to call `generation_allocator_efficiency_percent` with appropriate scaling, but I see that you were careful to avoid rounding errors due to integer truncation and I don't think we want to try to analyze if percent, per mille, etc., would be sufficient.

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -5105,20 +5105,24 @@ size_t& generation_allocated_since_last_pin (generation* inst)
+    uint64_t free_obj_space = generation_free_obj_space (inst);
+    uint64_t free_list_allocated = generation_free_list_allocated (inst);
+    uint64_t free_list_space = generation_free_list_space (inst);
+    if (free_obj_space==0)
```
```suggestion
    if ((free_list_allocated + free_obj_space) == 0)
```

markples replied:
I think the idea is that if free_obj_space is zero, then the computation will be zero regardless of the free_list_allocated value.  (unless one can be negative...)  It's subtle and perhaps not worth the microoptimization or worth a comment.  The same patterns occurs in the efficiency method.

grospelliergilles replied:
Yes if `free_obj_space` is zero then the result is always zero. The operands are of type `size_t` and so can not be negative.

Maoni0 replied:
yes, I got the idea. but this kind of microoptimization is not worth it at all. I'd much rather have the code be easier to understand.

grospelliergilles replied:
Thanks for your report. I will do as requested.

grospelliergilles committed:
Use 'size_t' instead of 'uint64_t' for return type.

Also add comment to explain why we use integer division instead of
floating point division.

grospelliergilles committed:
Improve readability of the test

jkotas commented on src/coreclr/gc/gc.cpp:
```diff
@@ -3418,7 +3418,7 @@ gc_heap::dt_high_frag_p (gc_tuning_point tp,
                 dprintf (GTC_LOG, ("h%d: gen%d, frag is %zd, alloc effi: %d%%, unusable frag is %zd, ratio is %d",
                     heap_number, gen_number, dd_fragmentation (dd),
-                    (int)(100*generation_allocator_efficiency (generation_of (gen_number))),
+                    (int)(generation_allocator_efficiency_percent (generation_of (gen_number))),
```
All places that call `generation_allocator_efficiency_percent` cast the result to int or uint32_t. Should the result be int or uint32_t to avoid these casts?

grospelliergilles replied:
Yes I can to that. I did not change it in previous commits because of flags in `printf` but I can do the change if you want. The result is always positive. Do you prefer `int` or `uint32_t` ?

jkotas replied:
@Maoni0 Do you have a preference?

grospelliergilles committed:
Remove cast from 'size_t' to 'int' when using 'generation_allocator_efficiency_percent'.

grospelliergilles commented:
> @grospelliergilles is this PR ready to review again?

I have fixed compile errors so I think it is OK.

mangod9 commented:
@Maoni0 can you please take another look?

Maoni0 reviewed (approved):
LGTM

grospelliergilles commented:
Thank you very much @mangod9 for accepting my PR.
